# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [2]:
# import numpy and pandas
import pandas as pd
import numpy as np
from scipy.stats import trim_mean, mode, skew, gaussian_kde, pearsonr, spearmanr, beta
from statsmodels.stats.weightstats import ztest as ztest

from scipy.stats import ttest_ind, norm, t
from scipy.stats import f_oneway
from scipy.stats import sem
from scipy import stats

# Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the City of Chicago. We will start by loading the dataset and examining its contents

In [3]:
# Run this code:
salaries = pd.read_csv('../data/Current_Employee_Names__Salaries__and_Position_Titles.csv')

Examine the `salaries` dataset using the `head` function below.

In [4]:
# Your code here
salaries.head()

,Name,Job Titles,Department,Full or Part-Time,Salary or Hourly,Typical Hours,Annual Salary,Hourly Rate
0,"AARON, JEFFERY M",SERGEANT,POLICE,F,Salary,NaN,101442.0,NaN
1,"AARON, KARINA",POLICE OFFICER (ASSIGNED AS DETECTIVE),POLICE,F,Salary,NaN,94122.0,NaN
2,"AARON, KIMBERLEI R",CHIEF CONTRACT EXPEDITER,GENERAL SERVICES,F,Salary,NaN,101592.0,NaN
3,"ABAD JR, VICENTE M",CIVIL ENGINEER IV,WATER MGMNT,F,Salary,NaN,110064.0,NaN
4,"ABASCAL, REECE E",TRAFFIC CONTROL AIDE-HOURLY,OEMC,P,Hourly,20.0,NaN,19.86


# Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of all hourly workers is significantly different from $30/hr. Import the correct one sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

In [5]:
# selecting the data of the salaries

hourly_salaries = salaries[salaries['Salary or Hourly'] == 'Hourly'][['Hourly Rate']].copy()
hourly_salaries.head()
data=hourly_salaries


In [6]:
# Your code here

t_statistic, p_value = stats.ttest_1samp(a=data, popmean=30)
print(t_statistic , p_value)

[20.61980579] [4.32302405e-92]


Interpretation: Since the p-value is much lower than the standard threshold of 0.05, you completely reject the hypothesis that the mean of your data is 30.

Conclusion: There is absolute statistical certainty that the mean of your data is not 30.

# Challenge 3 - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

The confidence interval is computed in SciPy using the `t.interval` function. You can read more about this function [here](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.t.html).

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level, number of rows - 1 for degrees of freedom, the mean of the sample for the location parameter and the standard error for the scale. The standard error can be computed using [this](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.sem.html) function in SciPy.

In [7]:
# Your code here
from scipy import stats

# Calculate the sample mean (location parameter)
sample_mean = data.mean()

# Calculate the degrees of freedom (number of rows - 1)
degrees_of_freedom = len(data) - 1

# Calculate the standard error of the mean 
standard_error = stats.sem(data)

# Construct the 95% Confidence Interval using the t-distribution
confidence_interval = stats.t.interval(confidence=0.95, 
                                       df=degrees_of_freedom, 
                                       loc=sample_mean, 
                                       scale=standard_error)

print(f"Sample mean: {sample_mean.iloc[0]:.2f}")
print(f"Degrees of freedom: {degrees_of_freedom}")
print(f"Standard error of the mean: {standard_error}")
print(f"95% Confidence Interval for the mean hourly wage: {(confidence_interval[0].item()):.2f} to {(confidence_interval[1].item()):.2f}")


Sample mean: 32.79
Degrees of freedom: 8021
Standard error of the mean: [0.13523686]
95% Confidence Interval for the mean hourly wage: 32.52 to 33.05


The interval is very small, which means the estimate is very precise. The mean of the sample is 32.79, which is above the value of 30 that we were testing against in the previous question. 
The confidence interval is for the population mean hourly wage and that the value 30 is outside the interval which supports rejecting the null hypothesis that the mean hourly wage is 30.


# Challenge 4 - Hypothesis Tests of Proportions

Another type of one sample test is a hypothesis test of proportions. In this test, we examine whether the proportion of a group in our sample is significantly different than a fraction. 

You can read more about one sample proportion tests [here](http://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/SAS/SAS6-CategoricalData/SAS6-CategoricalData2.html).

In the cell below, use the `proportions_ztest` function from `statsmodels` to perform a hypothesis test that will determine whether the number of hourly workers in the City of Chicago is significantly different from 25% at the 95% confidence level.

In [ ]:
# Your code here
from statsmodels.stats.proportion import proportions_ztest

# Count the number of hourly workers (successes)
hourly_count = len(salaries[salaries['Salary or Hourly'] == 'Hourly'])

# Count the total number of workers (observations)
total_observations = len(salaries)

# null hypothesis value (25%)
null_hyp_proportion = 0.25

# Z-test for proportions
z_stat, p_value = proportions_ztest(count=hourly_count , nobs=total_observations , value=null_hyp_proportion)


print(f"Z-statistic: {z_stat:.4f}")
print(f"P-value: {p_value:.4f}")

Z-statistic: -3.5100
P-value: 0.0004


P-value: 0.0004

Since it is less than 0.05 (for a 95% confidence level), there is sufficient statistical evidence to say that the proportion is not 25%.
We reject the null hypothesis at the 5% significance level because the p-value is below 0.05.